### PR TITLE
APM-209

### PR DESCRIPTION
**Pull request by cpjulia**
### Scope & Purpose

This is related to https://arangodb.atlassian.net/browse/APM-209.
Histogram displaying is now switched off by default. For displaying it, the new flag `histogram.generate` must be set to true. Its default value is false for compatibility with other versions and also for complying with the histogram not being displayed by default. 
If this flag is not set to true, but other histogram flags are addressed, e.g. `--histogram.interval-size 500`, everything will still run normally, but a warning message will be displayed saying that the histogram is switched off and setting this flag would not be of use.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backport for 3.9: 

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/806
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket number: https://arangodb.atlassian.net/browse/APM-209
- [ ] Design document: 

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: arangosh/Benchmark/BenchFeature.cpp [hunk 1]
```diff
@@ -251,6 +271,50 @@ void BenchFeature::updateStartCounter() { ++_started; }
...
+
+      builder.add("IntervalSize", VPackValue(threads[i]->_histogramIntervalSize));
+
+      for (auto time : res) {
+        builder.add(std::to_string(_percentiles[j]), VPackValue(time));
+        pp << "   " << std::right << std::fixed << std::setw(9)
+           << std::setprecision(4) << (time * 1000) << std::setw(0) << "ms";
+        j++;
+      }
+      pp << std::endl;
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:312]:
```suggestion
      pp << '\n';
```

File: arangosh/Benchmark/BenchFeature.cpp [hunk 2]
```diff
@@ -251,6 +271,50 @@ void BenchFeature::updateStartCounter() { ++_started; }
...
 int BenchFeature::getStartCounter() { return _started; }
 
+void BenchFeature::setupHistogram(std::stringstream& pp, VPackBuilder& builder) {
+  builder.add("histogram", VPackValue(VPackValueType::Object));
+  pp << "Interval/Percentile:";
+  for (auto percentile : _percentiles) {
+    pp << std::fixed << std::right << std::setw(13) << std::setprecision(2)
+        << percentile << "%";
+  }
+  pp << std::endl;
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:281]:
```suggestion
  pp << '\n';
```

File: CHANGELOG [hunk 3]
```diff
@@ -11,6 +21,7 @@ devel
 * Fixed potential undefined behavior in edge cache during cache migration tasks.
   There was a short window of time in which an already freed Table could be
   used by concurrently running edge lookups.
+>>>>>>> 774abcf00a0b95d7b955532da0f58cd7e93fee6e
```
Comment (gnusi) [inline CHANGELOG:24]: Merging artifact?

File: arangosh/Benchmark/BenchFeature.cpp [hunk 4]
```diff
@@ -482,10 +525,13 @@ void BenchFeature::start() {
...
 }
 
-bool BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult> results,
+void BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult> results,
                           BenchmarkStats const& stats,
                           std::string const& histogram, VPackBuilder& builder) {
-  std::cout << std::endl;
+
+  if (_generateHistogram) {
+    std::cout << histogram << std::endl;
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:533]:
```suggestion
    std::cout << histogram << '\n';
```

File: arangosh/Benchmark/BenchFeature.cpp [hunk 5]
```diff
@@ -432,42 +493,24 @@ void BenchFeature::start() {
...
-      builder->close();
-      pp << std::endl;
-    }
     threads.clear();
   }
+  if (_generateHistogram) {
+    builder.close();
+  }
 
   std::cout << std::endl;
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:504]:
```suggestion
  std::cout << '\n';
```

File: arangosh/Benchmark/BenchFeature.cpp [hunk 6]
```diff
@@ -556,31 +603,24 @@ bool BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult> res
...
-  std::cout << "Min request time: " << std::setprecision(6)
+  std::cout << "Min request time: " << std::setprecision(4)
             << (stats.min * 1000) << "ms" << std::endl
-            << "Avg request time: " << std::setprecision(6)
+            << "Avg request time: " << std::setprecision(4)
             << (stats.avg() * 1000) << "ms" << std::endl
-            << "Max request time: " << std::setprecision(6)
+            << "Max request time: " << std::setprecision(4)
             << (stats.max * 1000) << "ms" << std::endl
             << std::endl;
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:612]:
```suggestion
            << '\n';
```

File: arangosh/Benchmark/BenchFeature.h [hunk 7]
```diff
@@ -74,15 +75,19 @@ class BenchFeature final : public application_features::ApplicationFeature {
...
+  void validateOptions(std::shared_ptr<options::ProgramOptions>) override final;
+
   
   std::string const& customQuery() const { return _customQuery; }
   std::string const& customQueryFile() const { return _customQueryFile; }
 
  private:
   void status(std::string const& value);
-  bool report(ClientFeature&, std::vector<BenchRunResult>, arangobench::BenchmarkStats const& stats, std::string const& histogram, VPackBuilder& builder);
+  void report(ClientFeature&, std::vector<BenchRunResult>, arangobench::BenchmarkStats const& stats, std::string const& histogram, VPackBuilder& builder);
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.h:86]: why to copy `vector`?

File: arangosh/Benchmark/BenchFeature.cpp [hunk 8]
```diff
@@ -432,42 +493,24 @@ void BenchFeature::start() {
...
+
+  builder.close();
+
+  if (!_jsonReportFile.empty()) {
+    auto json = builder.toJson();
+    TRI_WriteFile(_jsonReportFile.c_str(), json.c_str(), json.length());
+  }
 
   if (!ok) {
     std::cout << "At least one of the runs produced failures!" << std::endl;
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:516]:
```suggestion
    std::cout << "At least one of the runs produced failures!" << '\n';
```

File: arangosh/Benchmark/BenchFeature.cpp [hunk 9]
```diff
@@ -482,10 +525,13 @@ void BenchFeature::start() {
   *_result = ret;
 }
 
-bool BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult> results,
+void BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult> results,
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:528]: why do we copy a vector?

File: arangosh/Benchmark/BenchFeature.h [hunk 10]
```diff
@@ -100,6 +105,7 @@ class BenchFeature final : public application_features::ApplicationFeature {
   bool _verbose;
   bool _quiet;
   bool _waitForSync;
+  bool _generateHistogram = false; // don't generate histogram by default
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.h:108]: let's be consistent and initialize it in constructor similar to other members

**Review by gnusi — CHANGES_REQUESTED** [review]

File: arangosh/Benchmark/BenchFeature.cpp [hunk 11]
```diff
@@ -239,6 +242,24 @@ void BenchFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
       "--verbose", "print out replies if the HTTP header indicates DB errors", false);
 }
 
+void BenchFeature::validateOptions(std::shared_ptr<ProgramOptions> options) {
+  if (options->processingResult().touched("--histogram.interval-size")) {
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:246]: why do we warn a user even if `--histogram.generate` is `true`

File: arangosh/Benchmark/BenchFeature.cpp [hunk 12]
```diff
@@ -333,19 +399,15 @@ void BenchFeature::start() {
...
   std::vector<std::unique_ptr<BenchmarkThread>> threads;
   bool ok = true;
   std::vector<BenchRunResult> results;
   std::stringstream pp;
-  pp << "Interval/Percentile:";
-  for (auto percentile : _percentiles) {
-    pp << std::fixed << std::right << std::setw(12) << std::setprecision(2)
-       << percentile << "%";
+  if (_generateHistogram) {
+    setupHistogram(pp, builder);
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:409]: it's really hard to follow who is responsible for forming a builder structure. 
I'd be in favor of at least moving `builder.add("histogram", VPackValue(VPackValueType::Object));` out of `setupHistogram`, so everything is handled in `BenchFeature::start()`

File: arangosh/Benchmark/BenchFeature.cpp [hunk 13]
```diff
@@ -556,31 +605,24 @@ bool BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult> res
   printResult(output, builder);
   builder.close();
 
-  std::cout << "Min request time: " << std::setprecision(6)
+  std::cout << "Min request time: " << std::setprecision(4)
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:608]: is making results less precise intentional? why 6 digits was bad?

File: arangosh/Benchmark/BenchFeature.cpp [hunk 14]
```diff
@@ -482,10 +526,13 @@ void BenchFeature::start() {
   *_result = ret;
 }
 
-bool BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult> results,
+void BenchFeature::report(ClientFeature& client, std::vector<BenchRunResult>& results,
```
Comment (gnusi) [inline arangosh/Benchmark/BenchFeature.cpp:529]: I'd be in favor of passing `std::vector<BenchRunResult>& results` as `std::vector<BenchRunResult> const& results` and sort results in a caller. In `BenchFeature::report` we can add an assertion that results are sorted, like `TRI_ASSERT(std::is_sorted(...))`

File: js/client/modules/@arangodb/testsuites/arangobench.js [hunk 15]
```diff
@@ -56,119 +56,141 @@ const testPaths = {
 // //////////////////////////////////////////////////////////////////////////////
```
Comment (gnusi) [inline js/client/modules/@arangodb/testsuites/arangobench.js:57]: are there any tests checking new functionality?

**Review by gnusi — CHANGES_REQUESTED** [review]

File: js/client/modules/@arangodb/testsuites/arangobench.js (hunk 15 above)
Comment (cpjulia) [reply js/client/modules/@arangodb/testsuites/arangobench.js:57]: it's because, at the moment, we don't have any infrastructure to test arangobench

File: arangosh/Benchmark/BenchFeature.cpp (hunk 12 above)
Comment (cpjulia) [reply arangosh/Benchmark/BenchFeature.cpp:409]: Addressed in latest commit

File: arangosh/Benchmark/BenchFeature.cpp (hunk 13 above)
Comment (cpjulia) [reply arangosh/Benchmark/BenchFeature.cpp:608]: We actually don't need more than 4 digits for handling the measurements in the histogram, it's less than milliseconds

File: arangosh/Benchmark/BenchFeature.cpp (hunk 14 above)
Comment (cpjulia) [reply arangosh/Benchmark/BenchFeature.cpp:529]: Addressed in latest commit, but can still have room for more optimization

**Review by gnusi — APPROVED** [review]
LGTM
